File: ruby/lib/runapi/suno/resources/separate_audio_stems.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      # Splits a track into individual instrument stems (vocals, drums, bass, guitar, etc.).
      class SeparateAudioStems
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/separate_audio_stems"
        RESPONSE_CLASS = Types::SeparateAudioStemsResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedSeparateAudioStemsResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          Validators.validate_separate_audio_stems!(params, self)
        end
      end
    end
  end
end

File: python/src/runapi/suno/resources/separate_audio_stems.py
"""Suno separate_audio_stems resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import CompletedSeparateAudioStemsResponse, SeparateAudioStemsResponse


class SeparateAudioStems(Resource):
    """Separate a track into stems."""

    ENDPOINT = "/api/v1/suno/separate_audio_stems"

    RESPONSE_CLASS = SeparateAudioStemsResponse
    COMPLETED_RESPONSE_CLASS = CompletedSeparateAudioStemsResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Split a track into stems and poll until it completes.

        Args:
            **params: stem separation parameters.

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a stem separation task and return immediately with an id.

        Args:
            **params: stem separation parameters.

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a stem separation task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["separate-audio-stems"], params)
